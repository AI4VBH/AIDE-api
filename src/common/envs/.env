PORT=5000

KEYCLOAK_URL='http://0.0.0.0:8443'
KEYCLOAK_MAIN_REALM='master'
KEYCLOAK_REALM='Demo-Realm'
KEYCLOAK_CLIENTID='nest-app'
KEYCLOAK_SECRET='83790b4f-48cd-4b6c-ac60-451a918be4b9'
KEYCLOAK_STATIC_ROLES='["admin", "clinician"]'

DATABASE_HOST='localhost'
DATABASE_NAME='nest_api'
DATABASE_USER='postgres'
DATABASE_PASSWORD='admin'
DATABASE_PORT=5432

ELASTIC_HOST=''
ELASTIC_PORT=''
ELASTIC_USE_SSL=''
ELASTIC_USERNAME=''
ELASTIC_PASSWORD=''
ELASTIC_INDEX=''

MONAI_API_HOST='https://localhost:8081/'
MIG_API_HOST='http://localhost:5010/'

MINIO_HOST='localhost'
MINIO_PORT=9001
MINIO_USE_SSL=false
MINIO_BUCKET='monaideploy'
MINIO_ACCESS_KEY='wOzmbchico52AMp3'
MINIO_SECRET_KEY='zyoT9SiMLiTCvNT0DS636FMiN4yi2258'
