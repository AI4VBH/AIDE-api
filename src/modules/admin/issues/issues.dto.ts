import { IMonaiPayload } from '../payloads/payload.interface';
import {
  MonaiWorkflowInstance,
  MonaiWorkflowTask,
} from '../workflowinstances/workflowinstances.interface';

export class IssueDto {
  task_id: string;
  status: string;
  workflow_name: string;
  patient_name: string;
  patient_id: string;
  execution_time: string;
  workflow_instance_id: string;
  execution_id: string;

  public static from(
    task: MonaiWorkflowTask,
    relativePayload: IMonaiPayload,
    relativeWorkflowInstance: MonaiWorkflowInstance,
  ) {
    const issue = new IssueDto();
    issue.task_id = task.task_id;
    issue.workflow_instance_id = task.workflow_instance_id;
    issue.execution_id = task.execution_id;
    issue.patient_id = relativePayload?.patient_details?.patient_id;
    issue.patient_name = relativePayload?.patient_details?.patient_name;
    issue.status = task.status;
    issue.execution_time = task.task_start_time;
    issue.workflow_name = relativeWorkflowInstance.workflow_name;
    return issue;
  }

  // TODO: Implement fromEntity method
  // public static fromEntity(entity: Issue) {
  //   return this.from({
  //     task_id: entity.id,
  //   });
  // }

  // TODO: Implement toEntity method
  // public toEntity(issue: Issue = null) {
  //   const it = new Issue();
  //   it.task_id = this.task_id;
  //   return it;
  // }
}
