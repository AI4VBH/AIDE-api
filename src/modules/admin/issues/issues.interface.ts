export interface Task {
  task_id: number;
  status: string;
  model_name: string;
  patient_name: string;
  patient_id: string;
  execution_time: string;
}
