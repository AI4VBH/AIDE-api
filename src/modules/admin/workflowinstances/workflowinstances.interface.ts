/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

export interface MonaiWorkflowInstance {
  id: string;
  ae_title: string;
  workflow_id: string;
  workflow_name: string;
  start_time: string;
  payload_id: string;
  status: string;
  bucket_id: string;
  acknowledged_workflow_errors: string;
  tasks: MonaiWorkflowTask[];
  input_metadata: [{ [name: string]: string }];
}

export interface MonaiWorkflowTask {
  execution_id: string;
  workflow_instance_id: string;
  previous_task_id: string;
  task_id: string;
  task_type: string;
  task_start_time: string;
  status: string;
  reason: string;
  acknowledged_task_errors: string;
  input_artifacts: [{ [name: string]: string }];
  output_artifacts: [{ [name: string]: string }];
  output_directory: string;
  result: [{ [name: string]: string }];
  next_timeout: string;
  timeout_interval: number;
  execution_stats?: [{ [name: string]: string }];
}
