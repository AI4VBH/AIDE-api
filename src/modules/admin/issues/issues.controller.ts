import { Controller, Get, Patch } from '@nestjs/common';
import { Issue } from './issues.interface';
import { IssuesService } from './issues.service';

@Controller('issues')
export class IssuesController {
  constructor(private readonly appService: IssuesService) {}

  @Get()
  getHello(): Issue[] {
    return this.appService.getIssues();
  }

  @Patch()
  dismissTask(): Issue {
    return this.appService.dismissTask();
  }
}
