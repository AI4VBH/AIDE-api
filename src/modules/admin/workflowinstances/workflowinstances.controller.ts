import { Controller, Put, Inject, Param, UseFilters } from '@nestjs/common';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import { WorkflowInstancesService } from './workflowinstances.service';

@Controller('workflowinstances')
@UseFilters(ExternalServerExceptionFilter)
export class WorkflowInstanceController {
  @Inject(WorkflowInstancesService)
  private readonly wfiService: WorkflowInstancesService;

  @Put(':workflow_instance_id/executions/:execution_id/acknowledge')
  acknowledgeTaskError(
    @Param('workflow_instance_id') workflow_instance_id,
    @Param('execution_id') execution_id,
  ) {
    return this.wfiService.acknowledgeTaskError(
      workflow_instance_id,
      execution_id,
    );
  }
}
