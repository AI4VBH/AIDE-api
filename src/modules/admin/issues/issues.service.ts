import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { AxiosResponse } from 'axios';
import { lastValueFrom } from 'rxjs';
import { IMonaiPayload } from '../payloads/payload.interface';
import { MonaiWorkflowInstance } from '../workflowinstances/workflowinstances.interface';
import { IssueDto } from './issues.dto';

@Injectable()
export class IssuesService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  async getIssues(
    workflowInstances: MonaiWorkflowInstance[],
  ): Promise<IssueDto[]> {
    const payloadIds = [
      ...new Set(workflowInstances.map((wf) => wf.payload_id)),
    ];

    const promises: Promise<AxiosResponse<IMonaiPayload>>[] = [];

    for (const payloadId of payloadIds) {
      promises.push(
        lastValueFrom(
          this.httpService.get<IMonaiPayload>(`payloads/${payloadId}`),
        ),
      );
    }

    const payloads = (await Promise.all(promises)).map((resp) => resp.data);

    return this.mapIssues(payloads, workflowInstances);
  }

  private mapIssues(
    payloads: IMonaiPayload[],
    workflowInstances: MonaiWorkflowInstance[],
  ): IssueDto[] {
    const issues: IssueDto[] = [];

    for (const workflowInstance of workflowInstances) {
      const relativePayload = payloads.find(
        (p) => p.payload_id == workflowInstance.payload_id,
      );

      const wfiIssues = workflowInstance.tasks.map((task) =>
        IssueDto.from(task, relativePayload, workflowInstance),
      );

      issues.push(...wfiIssues);
    }

    return issues;
  }
}
