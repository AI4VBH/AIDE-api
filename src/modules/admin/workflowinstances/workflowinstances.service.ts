import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { lastValueFrom } from 'rxjs';
import { MonaiWorkflowInstance } from './workflowinstances.interface';

@Injectable()
export class WorkflowInstancesService {
  @Inject(HttpService)
  private readonly httpService: HttpService;
  private readonly CONTROLLER: string = 'workflowinstances';
  async acknowledgeTaskError(
    workflow_instance_id: string,
    execution_id: string,
  ): Promise<MonaiWorkflowInstance> {
    const response = await lastValueFrom(
      this.httpService.put<MonaiWorkflowInstance>(
        `${this.CONTROLLER}/${workflow_instance_id}/executions/${execution_id}/acknowledge`,
      ),
    );

    return response.data;
  }

  async getAcknowledgedTaskErrors(): Promise<MonaiWorkflowInstance[]> {
    const response = await lastValueFrom(
      this.httpService.get<MonaiWorkflowInstance[]>('workflowinstances/failed'),
    );

    return response.data;
  }
}
