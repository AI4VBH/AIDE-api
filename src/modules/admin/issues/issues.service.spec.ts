/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { DeepMocked, createMock } from '@golevelup/ts-jest';
import { HttpService } from '@nestjs/axios';
import { Logger } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { AxiosInstance } from 'axios';
import { makeObservableForTest } from 'test/utilities/test-make-observable';
import { IMonaiPayload } from '../payloads/payload.interface';
import { MonaiWorkflowInstance } from '../workflowinstances/workflowinstances.interface';
import { IssuesService } from './issues.service';

describe('IssuesServices', () => {
  let axios: DeepMocked<AxiosInstance>;
  let httpService: DeepMocked<HttpService>;
  let service: IssuesService;

  beforeEach(async () => {
    axios = createMock<AxiosInstance>();
    httpService = createMock<HttpService>({
      axiosRef: axios,
    });

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        IssuesService,
        {
          provide: HttpService,
          useValue: httpService,
        },
        {
          provide: Logger,
          useFactory: () => createMock<Logger>(),
        },
      ],
    }).compile();

    service = module.get<IssuesService>(IssuesService);
  });

  it('should be defined', () => expect(service).toBeDefined());
  describe('get expected issues', () => {
    it('returns the expected result', async () => {
      const givenWorkflowInstances: MonaiWorkflowInstance[] = [
        {
          id: 'c5e980ff-e278-441b-af64-8630bf57b7e3',
          ae_title: 'some-title',
          payload_id: '041293d0-ab97-4ea1-b967-42ec62f26608',
          workflow_id: 'fca60a4d-2932-4196-9206-cf123c82332b',
          start_time: '2022-09-22T12:45:13.102Z',
          acknowledged_workflow_errors: '2022-09-22T12:45:13.102Z',
          status: 'Failed',
          tasks: [
            {
              task_id: 'email-task',
              workflow_instance_id: 'c5e980ff-e278-441b-af64-8630bf57b7e3',
              execution_id: '981eca65-294a-434a-930a-b1828b54253a',
              previous_task_id: 'export-task',
              task_start_time: '2022-09-23T12:50:00.305Z',
              task_type: 'test',
              status: 'Failed',
              acknowledged_task_errors: '2022-09-23T12:50:00.305Z',
              reason: 'undefined',
              input_artifacts: [{ name: '', value: '' }],
              output_artifacts: [{ name: '', value: '' }],
              output_directory: 'outputdir',
              result: [{ name: '', value: '' }],
              next_timeout: '2',
              timeout_interval: 60,
            },
            {
              task_id: 'email-task-2',
              workflow_instance_id: 'c5e980ff-e278-441b-af64-8630bf57b7e3',
              execution_id: '0000b113-0f60-42b2-b9fd-849ad209060c',
              previous_task_id: 'export-task',
              task_start_time: '2022-09-23T12:50:00.305Z',
              task_type: 'test',
              status: 'Succeeded',
              acknowledged_task_errors: '2022-09-23T12:50:00.305Z',
              reason: 'undefined',
              input_artifacts: [{ name: '', value: '' }],
              output_artifacts: [{ name: '', value: '' }],
              output_directory: 'outputdir',
              result: [{ name: '', value: '' }],
              next_timeout: '2',
              timeout_interval: 60,
            },
          ],
          bucket_id: 'bucket',
          input_metadata: [{ name: '', value: '' }],
          workflow_name: 'lillie',
        },
      ];

      const expectedPayload: IMonaiPayload = {
        id: 'Id',
        payload_id: '041293d0-ab97-4ea1-b967-42ec62f26608',
        workflows: [],
        workflow_instance_ids: ['c5e980ff-e278-441b-af64-8630bf57b7e3'],
        file_count: 0,
        correlation_id: 'd37b8de8-e265-504d-8ab3-bdbff199d206xx',
        bucket: 'bucket',
        calling_aetitle: 'B99E00',
        called_aetitle: '484406',
        timestamp: '2022-04-14T16:25:37+01:00',
        files: [],
        patient_details: {
          patient_id: '29FA59',
          patient_name: 'Rebecca',
          patient_sex: 'Franklin',
          patient_dob: '20/08/2022',
          patient_age: '25',
          patient_hospital_id: 'Buvveho',
        },
      };

      axios.get.mockResolvedValue({ data: expectedPayload });
      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const response = await service.getIssues(givenWorkflowInstances);

      expect(response).toMatchSnapshot();
    });
  });
});
