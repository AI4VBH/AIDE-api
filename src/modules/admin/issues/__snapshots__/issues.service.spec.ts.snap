// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssuesServices get expected issues returns the expected result 1`] = `
[
  IssueDto {
    "execution_id": "981eca65-294a-434a-930a-b1828b54253a",
    "execution_time": "2022-09-23T12:50:00.305Z",
    "patient_id": "29FA59",
    "patient_name": "Rebecca",
    "payload_id": "041293d0-ab97-4ea1-b967-42ec62f26608",
    "status": "Failed",
    "task_id": "email-task",
    "workflow_instance_id": "c5e980ff-e278-441b-af64-8630bf57b7e3",
    "workflow_name": "lillie",
  },
  IssueDto {
    "execution_id": "981eca65-294a-434a-930a-b1828b54253a",
    "execution_time": "2022-09-23T12:10:00.305Z",
    "patient_id": "29FA59",
    "patient_name": "Rebecca",
    "payload_id": "041293d0-ab97-4ea1-b967-42ec62f26608",
    "status": "Failed",
    "task_id": "email-task-3",
    "workflow_instance_id": "c5e980ff-e278-441b-af64-8630bf57b7e3",
    "workflow_name": "lillie",
  },
]
`;
